fsharp-backend/src/LibBackend/StaticAssets.fs

```diff
@@ -107,8 +97,132 @@ let latestDeployHash (canvasID : CanvasID) : Task<Option<string>> =
+  : Task<List<StaticDeploy>> =
+  Sql.query
+    "SELECT deploy_hash, created_at, live_at
+     FROM static_asset_deploys
+     WHERE canvas_id=@canvasID
+     ORDER BY created_at
+     DESC LIMIT 25"
+  |> Sql.parameters [ "canvasID", Sql.uuid canvasID ]
+  |> Sql.executeAsync (fun read ->
+    let deployHash = read.string "deploy_hash"
+
+    let status, lastUpdate =
+      match read.instantOrNone "live_at" with
+      | Some datetime -> Deployed, datetime
+      | None -> Deploying, read.instant "created_at"
+
+    { deployHash = deployHash
+      url = url canvasName deployHash Short
+      status = status
+      lastUpdate = lastUpdate })
+
+let startStaticAssetDeploy
+  (user : Account.UserInfo)
+  (canvasID : CanvasID)
+  (canvasName : CanvasName.T)
+  : Task<StaticDeploy> =
+
+  // we include .fff (milliseconds) to ensure we don't encoutner conflicts,
+  // especially relevant to unit tests which record multiple deploys quickly.
+  let now = System.DateTime.Now.ToString("MM/dd/yyyy hh:mm:ss.fff tt")
+
+  let deployHash =
+    $"{canvasID}{now}"
+    |> sha1digest
+    |> Base64.urlEncodeToString
+    |> String.removeSuffix "="
+    |> String.toLowercase
+    |> String.take 10
+
+  Sql.query
+    "INSERT INTO static_asset_deploys
+      (canvas_id, branch, deploy_hash, uploaded_by_account_id)
+    VALUES (@canvasID, @branch, @deployHash, @uploadedBy)
+    RETURNING created_at"
+  |> Sql.parameters [ "canvasID", Sql.uuid canvasID
+                      "branch", Sql.string branch
+                      "deployHash", Sql.string deployHash
+                      "uploadedBy", Sql.uuid user.id ]
+  |> Sql.executeRowAsync (fun reader -> reader.instant "created_at")
+  |> Task.map (fun lastUpdate ->
+    { deployHash = deployHash
+      url = url canvasName deployHash Short
+      lastUpdate = lastUpdate
+      status = Deploying })
+
+// CLEANUP: return an Error if the deploy hash doesn't exist
+// CLEANUP: decide what to do if the deploy is already finished
+let finishStaticAssetDeploy
+  (canvasID : CanvasID)
+  (canvasName : CanvasName.T)
+  (deployHash : string)
+  : Task<StaticDeploy> =
+  Sql.query
+    "UPDATE static_asset_deploys
+      SET live_at = NOW()
+      WHERE canvas_id = @canvasID AND deploy_hash = @deployHash
+      RETURNING live_at"
+  |> Sql.parameters [ "canvasID", Sql.uuid canvasID
+                      "deployHash", Sql.string deployHash ]
+  |> Sql.executeRowAsync (fun reader -> reader.instant "live_at")
+  |> Task.map (fun lastUpdate ->
+    { deployHash = deployHash
+      url = url canvasName deployHash Short
+      lastUpdate = lastUpdate
+      status = Deployed })
 
 
+/// <summary>Removes references to a canvas' static asset deploy</summary>
+/// <remarks>
+/// Since postgres doesn't have named transactions, we just delete the db
+/// record in question. For now, we're leaving files where they are; the right
+/// thing to do here would be to shell out to `gsutil -m rm -r` - leaving  this
```

Comment:
This comment is out of date. It's clear we should handle the deletion in the canvas.

Reply:
Enh, I'll revert on second thought - the description alone should make it clear.